crypto/x509/x509_vfy.c

```diff
@@ -2374,9 +2383,9 @@ int X509_STORE_CTX_purpose_inherit(X509_STORE_CTX *ctx, int def_purpose,
         }
     }
 
-    if (ctx->param->purpose == 0 && purpose != 0)
+    if ((ctx->param->purpose == 0 && purpose != 0) || override != 0)
         ctx->param->purpose = purpose;
-    if (ctx->param->trust == 0 && trust != 0)
+    if ((ctx->param->trust == 0 && trust != 0) || override != 0)
```

Comment:
To avoid the not really intended (nor documented) side effect on the purpose when only the trust shall be overridden, or on the trust when only the purpose shall be overridden, change this to:
```
    if ((ctx->param->purpose == 0 && purpose != 0)
            || override && purpose != 0)
        ctx->param->purpose = purpose;
    if ((ctx->param->trust == 0 && trust != 0)
            || override && trust != 0)
```